spec/blocksync/verification.md

```diff
@@ -0,0 +1,120 @@
+
+If the block received does not verify using the same verification logic, the peer is faulty. In case the verification passes, but headers mismatch, the node will not know who to trust.
+
+**Verification failed**
+
+In the light client verification model, the client requests from both peers a trace of blocks in order to determine the block where the divergence happened. As the blocks being verified are within the trusted period, the light client can generate evidence and send it to full nodes to vote for it and determine which peer is faulty. 
+
+But in the block sync reactor, if verification fails, there are multiple possible scenarios:
+
+- Abort blocksync and switch to consensus.
+- Request data from more peers.  
+
+Both cases can be a potential attack on the joining node. We discuss these different attack possibilities and design shortcomings in the section below.
+
+
+
+## Problems with current verification and reactor design
+
+
+The crux of many concerns with regards to the correctness and safety of the block sync reactor lies in the fact a node connects to peers. 
+
+Currently, the only constraint a node has is that it has to connect to **a** peer. An improved design should require a node to connect to at least two peers. 
+
+We have no guarantees on the correctness of the peers we are connected to. 
+- Once we replace a peer due to it having timed out or not reporting a correct block, nothing prevents a node from reconnecting to it - *there is no notion of blacklisted peers*.
+- If we connect to a subset of peers, they could feed the node faulty data. Eventually, when the node switches to consensus, it would realize there is something wrong, but then the node itself might be blacklisted.
+- Alternatively, a node that is fed faulty data, could, upon switching to conensus, become part of a light client attack by serving as a faulty witness. 
+- There is no check whether the maximum height reported by peers is true or not. A slow node could report a very distant height to the node - for example 2000, when the blockchain is at height 1000 in fact. This would lead to one part of the condition to switch to consensus never being true. To prevent the node switching due to not advancing, the malicious node sends a new block very slowly. Thus the node progresses but can never participate in consensus. This issue could potentially be mitigated if, instead of taking the maximum height reported by peers, we report the lowest of their maximums. The idea is that peers should be close enought to the top of the chain in any case. 
+- A blocksyncing node can flood peers with requests - constantly reporting that it has not synced up. At the moment the maximum amount of requests received is limited and protects peers to some extend against this attack. 
```

Comment:
I could be wrong on this, but another problem is that the blocks sent/received via block synch reactor are not chunked like in consensus. So a busy chain containing big blocks might cause performance problems. Also, a malicious node could try to feed big blocks to try to "choke" the receiver.

I am aware this affects performance rather than correctness

Reply:
That is why I suggested that maybe we send around headers and blocks only once verification passes. However, this just delays the problem. .